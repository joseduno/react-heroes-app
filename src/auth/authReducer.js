import { types } from "../types/types"

export const authReducer = (state={}, action) => {



    switch (action?.type) {
        
        case types.login:
            return {
                // No entiendo por que utilizar el operador spread
                ...action.payload,
                logged: true
            }

        case types.logout:
            return {
                logged: false
            }
    
        default:
            return state
        }
}