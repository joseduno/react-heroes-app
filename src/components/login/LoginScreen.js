import React, { useContext } from 'react'
import { AuthContext } from '../../auth/AuthContext'
import { types } from '../../types/types'

export const LoginScreen = ({history}) => {

    const {dispatch} = useContext(AuthContext)
    const lastPath = localStorage.getItem('lastPath') || '/'

    const handleLogin = () => {

        dispatch({
            type: types.login,
            payload: {name: 'jose'}
        })

        // history.push('/')
        history.replace(lastPath)
        /* 
        a diferencia del push, la funcion replace no guarda el historial,
        Si se accede desde login y se vuelve a la pagina anterior del navegador,
        la página del login no se mostrará
         */

    }

    return (
        <div className='container mt-5'>
            <h1>Login</h1>
            <hr />
            <button
                className='btn btn-primary'
                onClick={handleLogin}
            >
                Ingresar
            </button>
        </div>
    )
}
