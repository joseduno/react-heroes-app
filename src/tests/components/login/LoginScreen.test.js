import { mount } from "enzyme"
import { AuthContext } from "../../../auth/AuthContext"
import { LoginScreen } from "../../../components/login/LoginScreen"
import { types } from "../../../types/types"


describe('Pruebas en el componente <LoginScreen />', () => {

    const historyMock = {replace: jest.fn()}

    const contextValue = {
        user: {logged: false},
        dispatch: jest.fn()
    }

    Storage.prototype.getItem = jest.fn()

    const wrapper = mount(
        <AuthContext.Provider value={contextValue}>
            <LoginScreen history={historyMock} />
        </AuthContext.Provider>
    )

    const handleLogin = wrapper.find('button').prop('onClick')

    test('debe mostrarse correctamente con usuario no autenticado', () => {
        expect(wrapper).toMatchSnapshot()
    })

    test('debe llamar al dispatch y redireccionar a / sin valor en localStorage', () => {
        
        handleLogin()
        expect(contextValue.dispatch).toHaveBeenCalledWith({
            type: types.login,
            payload: expect.any(Object)
        })
        expect(historyMock.replace).toHaveBeenCalledWith('/')

    })
    
})
