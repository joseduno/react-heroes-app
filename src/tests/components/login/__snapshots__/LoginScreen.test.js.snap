// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente <LoginScreen /> debe mostrarse correctamente con usuario no autenticado 1`] = `
<div
  className="container mt-5"
>
  <h1>
    Login
  </h1>
  <hr />
  <button
    className="btn btn-primary"
    onClick={[Function]}
  >
    Ingresar
  </button>
</div>
`;
